File: №8 Bridge Pattern/Java/src/Main.java

interface MovieBooking {
    void bookTicket(int numberOfTickets);
    void cancelBooking();
}

interface MovieBookingSystem {
    void makeBooking(String movie, int numberOfTickets);
    void cancelBooking(String movie);
}

class OnlineBookingSystem implements MovieBookingSystem {
    @Override
    public void makeBooking(String movie, int numberOfTickets) {
        System.out.println("Online booking for " + numberOfTickets + " tickets for " + movie);
    }

    @Override
    public void cancelBooking(String movie) {
        System.out.println("Canceling online booking for " + movie);
    }
}

class BoxOfficeBookingSystem implements MovieBookingSystem {
    @Override
    public void makeBooking(String movie, int numberOfTickets) {
        System.out.println("Box office booking for " + numberOfTickets + " tickets for " + movie);
    }

    @Override
    public void cancelBooking(String movie) {
        System.out.println("Canceling box office booking for " + movie);
    }
}

class MovieBookingImpl implements MovieBooking {
    private MovieBookingSystem bookingSystem;
    private String movie;

    public MovieBookingImpl(MovieBookingSystem bookingSystem, String movie) {
        this.bookingSystem = bookingSystem;
        this.movie = movie;
    }

    @Override
    public void bookTicket(int numberOfTickets) {
        bookingSystem.makeBooking(movie, numberOfTickets);
    }

    @Override
    public void cancelBooking() {
        bookingSystem.cancelBooking(movie);
    }
}

public class Main {
    public static void main(String[] args) {
        MovieBookingSystem onlineBookingSystem = new OnlineBookingSystem();
        MovieBookingSystem boxOfficeBookingSystem = new BoxOfficeBookingSystem();

        MovieBooking onlineBooking = new MovieBookingImpl(onlineBookingSystem, "Пірати карибського моря");
        MovieBooking boxOfficeBooking = new MovieBookingImpl(boxOfficeBookingSystem, "Бетмен");

        onlineBooking.bookTicket(2);
        onlineBooking.cancelBooking();

        boxOfficeBooking.bookTicket(3);
        boxOfficeBooking.cancelBooking();
    }
}

File: №8 Bridge Pattern/Python/main.py
from abc import ABC, abstractmethod


class MovieBooking(ABC):
    @abstractmethod
    def book_ticket(self, number_of_tickets):
        pass

    @abstractmethod
    def cancel_booking(self):
        pass


class MovieBookingSystem(ABC):
    @abstractmethod
    def make_booking(self, movie, number_of_tickets):
        pass

    @abstractmethod
    def cancel_booking(self, movie):
        pass


class OnlineBookingSystem(MovieBookingSystem):
    def make_booking(self, movie, number_of_tickets):
        print(f"Online booking for {number_of_tickets} tickets for {movie}")

    def cancel_booking(self, movie):
        print(f"Canceling online booking for {movie}")


class BoxOfficeBookingSystem(MovieBookingSystem):
    def make_booking(self, movie, number_of_tickets):
        print(f"Box office booking for {number_of_tickets} tickets for {movie}")

    def cancel_booking(self, movie):
        print(f"Canceling box office booking for {movie}")


class MovieBookingImpl(MovieBooking):
    def __init__(self, booking_system, movie):
        self.booking_system = booking_system
        self.movie = movie

    def book_ticket(self, number_of_tickets):
        self.booking_system.make_booking(self.movie, number_of_tickets)

    def cancel_booking(self):
        self.booking_system.cancel_booking(self.movie)


if __name__ == "__main__":
    online_booking_system = OnlineBookingSystem()
    box_office_booking_system = BoxOfficeBookingSystem()

    online_booking = MovieBookingImpl(online_booking_system, "Пірати карибського моря")
    box_office_booking = MovieBookingImpl(box_office_booking_system, "Бетмен")

    online_booking.book_ticket(2)
    online_booking.cancel_booking()

    box_office_booking.book_ticket(3)
    box_office_booking.cancel_booking()
